### PR TITLE
feat(service-portal): Mileage

### PR DESCRIPTION
**Pull request by thordurhhh**
## What

Add vehicle mileage form. Form and page to add mileage reading.

## Why

New feature

## Screenshots / Gifs

Screenshot 2023-11-24 at 13 39 33
---
Screenshot 2023-11-24 at 13 39 26
---
Screenshot 2023-11-24 at 13 39 15

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: libs/api/domains/vehicles/src/lib/api-domains-vehicles-shared.resolver.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  })
+  async resolveLastMileage(
+    @Context('req') { user }: { user: User },
+    @Parent() overview: VehiclesDetail,
+  ): Promise<VehicleMileageDetail | null> {
+    const res = await this.vehiclesService.getVehicleMileage(user, {
+      permno: overview.basicInfo?.permno ?? '',
+    })
+
+    this.logger.debug('resolveLastMileage data', res?.data)
```
Comment (thorkellmani) [inline libs/api/domains/vehicles/src/lib/api-domains-vehicles-shared.resolver.ts:33]:
```suggestion
```

File: libs/api/domains/vehicles/src/lib/api-domains-vehicles.service.ts [hunk 2]
```diff
@@ -7,16 +7,41 @@ import {
...
+
+  return isSameDay(today, inputDate)
+}
 
 @Injectable()
 export class VehiclesService {
   constructor(
     @Inject(VehicleSearchApi) private vehiclesApi: VehicleSearchApi,
+    @Inject(MileageReadingApi) private mileageReadingApi: MileageReadingApi,
+    @Inject(PdfApi) private vehiclesPDFApi: PdfApi,
```
Comment (thorkellmani) [inline libs/api/domains/vehicles/src/lib/api-domains-vehicles.service.ts:44]:
```suggestion
    private vehiclesApi: VehicleSearchApi,
    private mileageReadingApi: MileageReadingApi,
    private vehiclesPDFApi: PdfApi,
```

File: libs/service-portal/assets/src/components/MileageConfirmation/index.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+          onClick={() => setModalIsOpen(true)}
+          loading={postActionLoading}
+        >
+          {formatMessage(m.save)}2
+        </Button>
+      }
+    >
+      <Box>
+        <Text>
+          Þú ert að fara að færa inn kílómetrastöðu: {formValue || '0 km'}
```
Comment (Tryggvig) [inline libs/service-portal/assets/src/components/MileageConfirmation/index.tsx:43]: formatmessage?

File: libs/service-portal/assets/src/components/MileageConfirmation/index.tsx [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      disclosure={
+        <Button
+          variant="primary"
+          size="small"
+          fluid
+          type="button"
+          onClick={() => setModalIsOpen(true)}
+          loading={postActionLoading}
+        >
+          {formatMessage(m.save)}2
```
Comment (Tryggvig) [inline libs/service-portal/assets/src/components/MileageConfirmation/index.tsx:37]: 2?

File: libs/api/domains/vehicles/src/models/getVehicleSearch.model.ts [hunk 5]
```diff
@@ -67,4 +67,13 @@ export class VehiclesVehicleSearch {
...
   weightedco2Wltp?: number
+
+  @Field(() => Date, { nullable: true })
+  nextAvailableMileageReadDate?: Date | null
+
+  @Field(() => Boolean, { nullable: true })
+  requiresMileageRegistration?: boolean | null
+
+  @Field(() => Boolean, { nullable: true })
+  canRegisterMileage?: boolean | null
```
Comment (thorkellmani) [inline libs/api/domains/vehicles/src/models/getVehicleSearch.model.ts:78]: Do we need both?

File: libs/api/domains/vehicles/src/models/usersVehicles.model.ts [hunk 6]
```diff
@@ -108,6 +108,15 @@ export class VehiclesVehicle {
...
   nextInspectionDate?: Date
+
+  @Field(() => Date, { nullable: true })
+  nextAvailableMileageReadDate?: Date | null
+
+  @Field(() => Boolean, { nullable: true })
+  requiresMileageRegistration?: boolean | null
+
+  @Field(() => Boolean, { nullable: true })
+  canRegisterMileage?: boolean | null
```
Comment (thorkellmani) [inline libs/api/domains/vehicles/src/models/usersVehicles.model.ts:119]: Do we need both?

File: libs/clients/vehicles-mileage/src/lib/vehiclesMileageClient.config.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    schema,
+    load(env) {
+      return {
+        xRoadServicePath: env.required(
+          'XROAD_VEHICLES_MILEAGE_PATH',
+          'IS-DEV/GOV/10017/Samgongustofa-Protected/Vehicle-Mileagereading-V1',
+        ),
+        fetch: {
+          timeout: 30000,
+        },
```
Comment (thorkellmani) [inline libs/clients/vehicles-mileage/src/lib/vehiclesMileageClient.config.ts:25]: Is there a purpose for elongating the timeout?

File: libs/service-portal/assets/src/screens/Overview/Overview.tsx [hunk 8]
```diff
@@ -155,6 +155,7 @@ const VehiclesOverview = () => {
       />
     )
   }
+
```
Comment (Tryggvig) [inline libs/service-portal/assets/src/screens/Overview/Overview.tsx:158]: boop

File: libs/clients/vehicles-mileage/src/lib/vehiclesMileageClient.service.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    xroadConfig: ConfigType<typeof XRoadConfig>,
+    config: ConfigType<typeof VehiclesMileageClientConfig>,
+    idsClientConfig: ConfigType<typeof IdsClientConfig>,
+  ) =>
+    new MileageReadingApi(
+      new Configuration({
+        fetchApi: createEnhancedFetch({
+          name: 'clients-mileage-vehicles',
+          organizationSlug: 'samgongustofa',
+          timeout: config.fetch.timeout,
```
Comment (thorkellmani) [inline libs/clients/vehicles-mileage/src/lib/vehiclesMileageClient.service.ts:25]: Why change timeout from default?

File: libs/service-portal/assets/src/components/MileageConfirmation/index.tsx [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+import { useLocale, useNamespaces } from '@island.is/localization'
+import { m, Modal } from '@island.is/service-portal/core'
+
+interface Props {
+  formValue?: string
+  postActionLoading?: boolean
+  buttonRef: React.RefObject<HTMLButtonElement>
+}
+
+const MileageConfirmation: FC<React.PropsWithChildren<Props>> = ({
```
Comment (thorkellmani) [inline libs/service-portal/assets/src/components/MileageConfirmation/index.tsx:13]: Is this used?

File: libs/service-portal/assets/src/lib/navigation.ts [hunk 11]
```diff
@@ -27,12 +27,20 @@ export const assetsNavigation: PortalNavigationItem = {
...
+          breadcrumbHide: true,
           children: [
             {
               // Path param reference
               name: 'id',
               navHide: true,
               path: AssetsPaths.AssetsVehiclesDetail,
+              children: [
+                {
+                  name: m.vehicleMilage,
```
Comment (thorkellmani) [inline libs/service-portal/assets/src/lib/navigation.ts:39]: Milage u say? 🇮🇹

File: libs/service-portal/assets/src/lib/navigation.ts [hunk 12]
```diff
@@ -27,12 +27,20 @@ export const assetsNavigation: PortalNavigationItem = {
...
+                  navHide: true,
+                  path: AssetsPaths.AssetsVehiclesDetailMilage,
```
Comment (thorkellmani) [inline libs/service-portal/assets/src/lib/navigation.ts:41]: Milage u say? 🇮🇹

File: libs/service-portal/assets/src/lib/paths.ts [hunk 13]
```diff
@@ -5,6 +5,7 @@ export enum AssetsPaths {
   AssetsVehicles = '/eignir/okutaeki',
   AssetsMyVehicles = '/eignir/okutaeki/min-okutaeki',
   AssetsVehiclesDetail = '/eignir/okutaeki/min-okutaeki/:id',
+  AssetsVehiclesDetailMilage = '/eignir/okutaeki/min-okutaeki/:id/kilometrastada',
```
Comment (thorkellmani) [inline libs/service-portal/assets/src/lib/paths.ts:8]: Milage u say? 🇮🇹

File: libs/service-portal/assets/src/module.tsx [hunk 14]
```diff
@@ -29,6 +29,10 @@ const WorkMachinesDetail = lazy(() =>
   import('./screens/WorkMachinesDetail/WorkMachinesDetail'),
 )
 
+const VehicleMilage = lazy(() =>
+  import('./screens/VehicleMilage/VehicleMilage'),
```
Comment (thorkellmani) [inline libs/service-portal/assets/src/module.tsx:33]: Milage u say? 🇮🇹

File: libs/service-portal/assets/src/module.tsx [hunk 15]
```diff
@@ -96,6 +100,13 @@ export const assetsModule: PortalModule = {
         enabled: userInfo.scopes.includes(ApiScope.vehicles),
         element: <VehicleHistory />,
       },
+      {
+        name: m.vehicleMilage,
+        path: AssetsPaths.AssetsVehiclesDetailMilage,
```
Comment (thorkellmani) [inline libs/service-portal/assets/src/module.tsx:105]: Milage u say? 🇮🇹

File: libs/service-portal/assets/src/module.tsx [hunk 16]
```diff
@@ -96,6 +100,13 @@ export const assetsModule: PortalModule = {
...
+        enabled: userInfo.scopes.includes(ApiScope.vehicles),
+        key: 'VehicleMileage',
+        element: <VehicleMilage />,
```
Comment (thorkellmani) [inline libs/service-portal/assets/src/module.tsx:108]: Milage u say? 🇮🇹

File: libs/service-portal/assets/src/screens/Overview/Overview.tsx (hunk 8 above)
Comment (thorkellmani) [inline libs/service-portal/assets/src/screens/Overview/Overview.tsx:158]: cool

File: libs/service-portal/assets/src/screens/VehicleDetail/VehicleDetail.tsx [hunk 17]
```diff
@@ -284,29 +192,58 @@ const VehicleDetail = () => {
...
+                      icon="pencil"
+                      iconType="outline"
+                      size="default"
+                      type="button"
+                      variant="utility"
+                    >
+                      <LinkResolver
+                        href={
+                          id
+                            ? AssetsPaths.AssetsVehiclesDetailMilage.replace(
```
Comment (thorkellmani) [inline libs/service-portal/assets/src/screens/VehicleDetail/VehicleDetail.tsx:212]: Milage u say? 🇮🇹

File: libs/service-portal/assets/src/screens/VehicleMilage/VehicleMilage.tsx [hunk 18]
```diff
@@ -0,0 +1,338 @@
...
+                    }}
+                    label={formatMessage(messages.vehicleMilageInputLabel)}
+                    placeholder={formatMessage(
+                      messages.vehicleMilageInputPlaceholder,
+                    )}
+                  />
+                </GridColumn>
+                <GridColumn
+                  span={['1/1', '7/9', '6/9', '5/9', '2/9']}
+                  offset={['0', '0', '0', '0', '0']}
```
Comment (Tryggvig) [inline libs/service-portal/assets/src/screens/VehicleMilage/VehicleMilage.tsx:245]: ['0'] ?

File: libs/service-portal/assets/src/screens/VehicleMilage/VehicleMilage.tsx [hunk 19]
```diff
@@ -0,0 +1,338 @@
...
+  const [putAction, { loading: putActionLoading }] =
+    usePutVehicleMileageMutation({
+      onError: () => {
+        toast.error(formatMessage(m.errorTitle))
+      },
+      onCompleted: () => {
+        // Clear form, state and errors on success.
+        setFormValue('')
+        reset()
+        refetch()
```
Comment (thorkellmani) [inline libs/service-portal/assets/src/screens/VehicleMilage/VehicleMilage.tsx:99]: Make a clear function or something?

**Review by Tryggvig — CHANGES_REQUESTED** [review]

File: libs/service-portal/core/src/lib/messages.ts [hunk 20]
```diff
@@ -270,6 +270,10 @@ export const m = defineMessages({
     id: 'service.portal:my-vehicles',
     defaultMessage: 'Ökutækin mín',
   },
+  vehicleMilage: {
+    id: 'service.portal:my-vehicles-milage',
```
Comment (thorkellmani) [inline libs/service-portal/core/src/lib/messages.ts:274]: Milage u say? 🇮🇹

**Review by thorkellmani — CHANGES_REQUESTED** [review]
COOL, small issues tho

File: libs/service-portal/assets/src/screens/VehicleMilage/VehicleMilage.tsx (hunk 18 above)
Comment (thordurhhh) [reply libs/service-portal/assets/src/screens/VehicleMilage/VehicleMilage.tsx:245]: lol

File: libs/api/domains/vehicles/src/models/usersVehicles.model.ts (hunk 6 above)
Comment (thordurhhh) [reply libs/api/domains/vehicles/src/models/usersVehicles.model.ts:119]: yes

File: libs/api/domains/vehicles/src/models/getVehicleSearch.model.ts (hunk 5 above)
Comment (thordurhhh) [reply libs/api/domains/vehicles/src/models/getVehicleSearch.model.ts:78]: yes

**Review by Thorrik58 — APPROVED** [review]

**Review by Tryggvig — APPROVED** [review]
LGTM

File: libs/clients/vehicles-mileage/src/lib/vehiclesMileageClient.service.ts [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+  scope: LazyDuringDevScope,
+  useFactory: (
+    xroadConfig: ConfigType<typeof XRoadConfig>,
+    config: ConfigType<typeof VehiclesMileageClientConfig>,
+    idsClientConfig: ConfigType<typeof IdsClientConfig>,
+  ) =>
+    new MileageReadingApi(
+      new Configuration({
+        fetchApi: createEnhancedFetch({
+          name: 'clients-mileage-vehicles',
```
Comment (saevarma) [inline libs/clients/vehicles-mileage/src/lib/vehiclesMileageClient.service.ts:23]: ~ Just a nit pick but just match the library name
```suggestion
          name: 'clients-vehicles-mileage',
```

File: libs/api/domains/vehicles/src/lib/api-domains-vehicles-mileage.resolver.ts [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+  PutVehicleMileageInput,
+} from '../dto/postVehicleMileageInput'
+
+@UseGuards(IdsUserGuard, ScopesGuard)
+@Resolver(() => VehicleMileageOverview)
+@Audit({ namespace: '@island.is/api/vehicles' })
+export class VehiclesMileageResolver {
+  constructor(private readonly vehiclesService: VehiclesService) {}
+
+  @Scopes(ApiScope.vehicles)
```
Comment (saevarma) [inline libs/api/domains/vehicles/src/lib/api-domains-vehicles-mileage.resolver.ts:38]: ~ The `@Scopes` decorator can be on the class level so all methods are protected by it by default.

File: libs/api/domains/vehicles/src/lib/api-domains-vehicles-mileage.resolver.ts [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+    name: 'vehicleMileageDetails',
+    nullable: true,
+  })
+  @Audit()
+  async getVehicleMileage(
+    @Args('input') input: GetVehicleMileageInput,
+    @CurrentUser() user: User,
+  ) {
+    return await this.vehiclesService.getVehicleMileage(user, input)
+  }
```
Comment (saevarma) [inline libs/api/domains/vehicles/src/lib/api-domains-vehicles-mileage.resolver.ts:49]: ~ FYI async/await is unnecessary when returnin the promise from the service function. `return await` is only required inside a `try/catch` when you want to await the response to catch an error.
```suggestion
  getVehicleMileage(
    @Args('input') input: GetVehicleMileageInput,
    @CurrentUser() user: User,
  ) {
    return this.vehiclesService.getVehicleMileage(user, input)
  }
```

**Review by saevarma — APPROVED** [review]
LGTM 🏎️

**Review by brynjarorng — APPROVED** [review]
LGTM

**Pull request merged**